src/chainedvector.jl

```diff
@@ -186,19 +186,31 @@ Base.@propagate_inbounds function Base.setindex!(A::ChainedVector, v, i::Integer
 end
 
 # custom index type used in eachindex
-struct ChainedVectorIndex{A}
+struct ChainedVectorIndex{A} <: Integer
     array::A
+    array_i::Int
     i::Int
 end
 
-@inline Base.getindex(x::ChainedVectorIndex) = @inbounds x.array[x.i]
+import Base: +, -, *, <, >, <=, >=
+for f in (:+, :-, :*, :<, :>, :<=, :>=)
+    @eval $f(a::ChainedVectorIndex, b::Integer) = $f(a.i, b)
+    @eval $f(a::Integer, b::ChainedVectorIndex) = $f(a, b.i)
+end
+Base.convert(::Type{T}, x::ChainedVectorIndex) where {T <: Union{Signed, Unsigned}} = x.i
```

Comment:
```suggestion
Base.convert(::Type{T}, x::ChainedVectorIndex) where {T <: Union{Signed, Unsigned}} = convert(T, x.i)
```
The `Union{Signed, Unsigned}` is a non-standard restriction, but I agree with it - we should not do automatic conversion to other numeric types as it is error prone.